### ISSUE TITLE
Affiliate Products Checkout URL is completely malformed.

### ISSUE THREAD

**Issue by Pottypotsworth** [facebookincubator/facebook-for-woocommerce#327]
The title pretty much sums up this issue but I think that there is an error in the feed file which is causing all of this to happen, here is what I know...

1) Configuring shop inside Woo Commerce works fine, 350 products are identified but during product sync only a single product is uploaded.
2) Product sets exist that I cannot delete, they both have one product and one of them seems malformed...
![screen shot 2018-03-15 at 00 24 28](https://user-images.githubusercontent.com/20905543/37415813-4854e264-27e7-11e8-9b1b-6f4068199a74.png)
2) 353 products are rejected from the data source (yet only 350 were reported as found)...
![screen shot 2018-03-15 at 00 25 18](https://user-images.githubusercontent.com/20905543/37415867-64e0dd48-27e7-11e8-963a-b4643164785d.png)
3) The issue seems to be related to admin-ajax.php?
![screen shot 2018-03-15 at 00 25 54](https://user-images.githubusercontent.com/20905543/37415897-7cd042c2-27e7-11e8-9ff2-0661173ab982.png)
4) Trying to access the site will then results in a 503 error, and the log files show...
`

2018-03-14 23:42:03 | Error | 203.86.238.179 | 503 | POST /wp-admin/admin-ajax.php?nonce=410ffcc2fe HTTP/1.1 |   |   | 4.36 K | Apache SSL/TLS access
-- | -- | -- | -- | -- | -- | -- | -- | --
2018-03-14 23:43:03 | Error | 203.86.238.179 |   | (70007)The  timeout specified has expired:  AH01075: Error dispatching request to  :, referer:  https://www.prettyme.ph/wp-admin/admin-ajax.php?nonce=410ffcc2fe |   |   |   | Apache error
2018-03-14 23:42:05 | Error | 112.205.251.59 | 503 | POST /wp-admin/admin-ajax.php HTTP/1.1 |   |   | 1.53 K | Apache SSL/TLS access
2018-03-14 23:43:05 | Error | 112.205.251.59 |   | (70007)The  timeout specified has expired:  AH01075: Error dispatching request to  :, referer:  https://www.prettyme.ph/wp-admin/admin.php?page=wc-settings&tab=integration |   |   |   | Apache error
2018-03-14 23:42:15 | Error | 112.205.251.59 | 503 | POST /wp-admin/admin-ajax.php HTTP/1.1 |   |   | 1.53 K | Apache SSL/TLS access
2018-03-14 23:43:15 | Error | 112.205.251.59 |   | (70007)The  timeout specified has expired:  AH01075: Error dispatching request to  :, referer:  https://www.prettyme.ph/wp-admin/admin.php?page=wc-settings&tab=integration |   |   |   | Apache error
2018-03-14 23:42:16 | Error | 112.205.251.59 | 503 | GET / HTTP/1.1 |   |   | 1.53 K | Apache SSL/TLS access
2018-03-14 23:43:16 | Error | 112.205.251.59 |   | (70007)The  timeout specified has expired:  AH01075: Error dispatching request to  :, referer:  https://www.prettyme.ph/wp-admin/admin.php?page=wc-settings&tab=integration |   |   |   | Apache error
2018-03-14 23:42:21 | Error | 112.205.251.59 | 503 | POST /wp-admin/admin-ajax.php HTTP/1.1 |   |   | 1.53 K | Apache SSL/TLS access
2018-03-14 23:43:21 | Error | 112.205.251.59 |   | (70007)The  timeout specified has expired:  AH01075: Error dispatching request to  :, referer:  https://www.prettyme.ph/wp-admin/admin.php?page=wc-settings&tab=integration |   |   |   | Apache error
2018-03-14 23:42:25 | Error | 203.86.238.179 | 503 | POST /wp-admin/admin-ajax.php?nonce=410ffcc2fe HTTP/1.1 |   |   | 4.36 K | Apache SSL/TLS access
2018-03-14 23:43:25 | Error | 203.86.238.179 |   | (70007)The  timeout specified has expired:  AH01075: Error dispatching request to  :, referer:  https://www.prettyme.ph/wp-admin/admin-ajax.php?nonce=410ffcc2fe |   |   |   | Apache error
2018-03-14 23:42:27 | Error | 112.205.251.59 | 503 | POST /wp-admin/admin-ajax.php HTTP/1.1 |   |   | 1.39 K | Apache SSL/TLS access
2018-03-14 23:43:27 | Error | 112.205.251.59 |   | (70007)The  timeout specified has expired:  AH01075: Error dispatching request to  :, referer:  https://www.prettyme.ph/wp-admin/admin.php?page=wc-settings&tab=integration |   |   |   | Apache error

`

I am at a complete loss here as to how to fix this. Any ideas?

**Comment by dmitridr** [comment]
Hello,

Thanks for notifying us of this. This is really weird and bizarre. I recommend downgrading to 1.7.11 temporarily while we try to figure out what happened here. Doing so will also help us figure out if the problem was caused by the new feed upload approach or it's a conflict with another extension you may have.

Can you give us an ID of the page/catalog/anything so we can dig into it?

cc @mengyingdu this is really weird.

**Comment by Pottypotsworth** [comment]
Hey there, 

So i deleted 1.18 plugin and installed 1.17. Initially it seemed to work a little better as setup was much faster, however as soon as the product sync started the server 503 server error'ed again.

I think this is an error in my log - (70007)The timeout specified has expired: AH01075: Error dispatching request to :, referer: https://www.prettyme.ph/wp-admin/admin.php?page=wc-settings&tab=integration

As far as I can tell no products have been passed through to the Facebook store.

This is the Page - https://www.facebook.com/PrettyMePhilippines/ - let me know if you need further info.

Any further ideas?

Thanks

**Comment by Pottypotsworth** [comment]
It might be worth mentioning that all my products are affiliate products with custom fields. Not sure if that is messing things up? I am confirming that site seems to crash into 503 BEFORE any sync has happened with Facebook product manager (https://www.facebook.com/products/). At least with 1.18 it managed to sync that first product :p

**Comment by dmitridr** [comment]
Thanks for mentioning that and confirming on 1.7.11. We only added the data source in 1.8.0, which is why you see no data source there, the products are uploaded one by one in older versions.

That they are affiliate products could very well be related, but it can't be the only reason. I double checked just now and affiliate products upload just fine on our test servers. 

In particular the problem I see with your products is:
(1) The product link is wrong.
(2) The title is blank.
(3) The description is blank.

And this is the case for all products. Do you have any custom extensions that could be affecting these fields? I'm not sure why they would be blank. 

Could you also share with us the configuration for one of the products that failed to upload so we can try to set up a similar product on our test servers?

------------

The 503 seems like a separate problem. Does the 503 persist forever afterward, only for a short period? Does it affect only the backend or the frontend too? And how do you recover from it?

**Comment by Pottypotsworth** [comment]
Hey @dmitridr,

Thanks for your continued help with this, let me try and answer your questions and note anything else I feel is relevant...

1) The one single product that got uploaded in v 1.18 is https://www.prettyme.ph/buy-womens/cyber-waterproof-raincoat/ - you should be able to see from there that there is both a title and description, and as far as I am aware all my custom fields have been setup in the correct manner. There is some custom code in product-single.php in the way that these products are displayed, but there shouldn't be anything weird in the SQL Database.
2) When looking at the WooCommerce products page in WP-Admin, the "Facebook Shop" column says that a mishmash of live products are live in the shop. This clearly isn't the case as the FB shop is completely empty.
3) Here is a list of all active plugins in normal use. I actually tried to disable all plugins except Woocommerce & FB Commernce and try again, what happens this time is that there is NO 503 error BUT still no products are actually sync'd despite everything indicating that they were.
![screen shot 2018-03-16 at 15 49 30](https://user-images.githubusercontent.com/20905543/37509816-e74e2a86-2933-11e8-84cf-c11e074c15ad.png)
4) I STILL have two empty product sets (cannot be deleted) on https://www.facebook.com/products/catalogs/557420227975712/sets with one of them being that weird malformed one that I posted about above.
5) The 503 seems to be tied to another plugin, as it didn't happen with all plugins disabled as per point 2). But I don't yet know which one. The 503 affects the entire site, it last for about 15 minutes and seems to kick its self out of it.
6) I don't think it is important, but during Facebook WooCommerce setup it ALWAYS says a Facebook pixel is already installed on the site. This was the case before but I have been through the code line by line and there is no reference at all to the Facebook Pixel, so I have no idea why it claims there is one installed.

Thanks, just let me know what else you need to test.

**Comment by dmitridr** [comment]
Hey,

Thanks for going through and testing some more, that's really helpful! I discovered something in our logs when I believe you were testing on 1.7.x and I discovered something which I think may be a bug which only happens for affiliate products.

It appears that in 1.7.x the products are uploading *almost* correctly, but the checkout url is being set to something wrong for all products, but the checkout url is not strictly necessary for shops to work correctly.

Could you try using the wordpress plugin editor (`Plugins > Edit`) and finding [this line of code](https://github.com/facebookincubator/facebook-for-woocommerce/blob/v1.8.1/includes/fbproduct.php#L437)

And changing
`if ($checkout_url) {`

To
`if (false) {`

Then re-setting the plugin (using `Reconfigure > Advanced Options > Delete`) and seeing if the products upload? Doing `Force Resync` might also work. 

If this works, we might issue a related fix, but I think 1.8.x may have a different problem, and the malformed product set is still a mystery to me.

**Comment by Pottypotsworth** [comment]
Hey @dmitridr 

So that simple one line change seems to have done the trick, 350 products have sync'd from what i can tell initially) correctly and are just waiting for approval from Facebook so i can check the click through's and what have you.

All 350 products were placed into the "Store collection all products" set, and also the malformed set.

So now all that is left to do is think of a way we can put this through the FB bug bounty program and we've got some cash to spend on ads :p 

In all seriousness though, thanks so much for your help. Just drop me a message if you need someone to test 1.18 once you get round to making some changes etc.

Thank again.

**Comment by Pottypotsworth** [comment]
Hi @dmitridr - I just upgraded the plugin to the latest version, 1.8.3 and found that the same issue seems to be occurring. I had to force an update as new products weren't being copied across and now my site is 503'd. Was the "bug" in the checkout URL for affiliate products not patched?

Thanks

**Comment by dmitridr** [comment]
Hi @Pottypotsworth 

Yes, apologies we haven't patched it yet. You'll have to manually apply the earlier fix of changing the code. I'll follow up and create a quick patch for the next version so that when you update you won't have this issue again.
The product set issue is pretty bizarre, but likely harmless. Let us know if it happens again after your next update.

**Comment by dmitridr** [comment]
I'll close out this issue when the patch is ready.

**Comment by Pottypotsworth** [comment]
Understood, @dmitridr - I did look for that line of code but it seemed like the entire file had been modified in the latest version and the single line was not where it should be. Admittedly I didn't look very hard though so I will double check. I can wait until it is officially patched anyway, so no rush. Thanks.
